# Info


This project is used to practice and learn about classification


This project will demonstrates 5 most-popular classification showdown: KNNClassfier, GaussianNB, RandomForestClassifier, DecisionTreeClassifier, SVC


## Preparation


In [113]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import LabelEncoder, StandardScaler
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.metrics import accuracy_score
from sklearn.decomposition import PCA

In [114]:
train = pd.read_csv("train.csv")
test = pd.read_csv("test.csv")
train

id                species   margin1   margin2   margin3   margin4  \
0       1            Acer_Opalus  0.007812  0.023438  0.023438  0.003906   
1       2  Pterocarya_Stenoptera  0.005859  0.000000  0.031250  0.015625   
2       3   Quercus_Hartwissiana  0.005859  0.009766  0.019531  0.007812   
3       5        Tilia_Tomentosa  0.000000  0.003906  0.023438  0.005859   
4       6     Quercus_Variabilis  0.005859  0.003906  0.048828  0.009766   
..    ...                    ...       ...       ...       ...       ...   
985  1575   Magnolia_Salicifolia  0.060547  0.119140  0.007812  0.003906   
986  1578            Acer_Pictum  0.001953  0.003906  0.021484  0.107420   
987  1581     Alnus_Maximowiczii  0.001953  0.003906  0.000000  0.021484   
988  1582          Quercus_Rubra  0.000000  0.000000  0.046875  0.056641   
989  1584         Quercus_Afares  0.023438  0.019531  0.031250  0.015625   

      margin5   margin6   margin7  margin8  ...  texture55  texture56  \
0    0.011719  0.009766  0.027344      0.0  ...   0.007812   0.000000   
1    0.025391  0.001953  0.019531      0.0  ...   0.000977   0.000000   
2    0.003906  0.005859  0.068359      0.0  ...   0.154300   0.000000   
3    0.021484  0.019531  0.023438      0.0  ...   0.000000   0.000977   
4    0.013672  0.015625  0.005859      0.0  ...   0.096680   0.000000   
..        ...       ...       ...      ...  ...        ...        ...   
985  0.000000  0.148440  0.017578      0.0  ...   0.242190   0.000000   
986  0.001953  0.000000  0.000000      0.0  ...   0.170900   0.000000   
987  0.078125  0.003906  0.007812      0.0  ...   0.004883   0.000977   
988  0.009766  0.000000  0.000000      0.0  ...   0.083008   0.030273   
989  0.005859  0.019531  0.035156      0.0  ...   0.000000   0.000000   

     texture57  texture58  texture59  texture60  texture61  texture62  \
0     0.002930   0.002930   0.035156   0.000000   0.000000   0.004883   
1     0.000000   0.000977   0.023438   0.000000   0.000000   0.000977   
2     0.005859   0.000977   0.007812   0.000000   0.000000   0.000000   
3     0.000000   0.000000   0.020508   0.000000   0.000000   0.017578   
4     0.021484   0.000000   0.000000   0.000000   0.000000   0.000000   
..         ...        ...        ...        ...        ...        ...   
985   0.034180   0.000000   0.010742   0.000000   0.000000   0.000000   
986   0.018555   0.000000   0.011719   0.000000   0.000000   0.000977   
987   0.004883   0.027344   0.016602   0.007812   0.000000   0.027344   
988   0.000977   0.002930   0.014648   0.000000   0.041992   0.000000   
989   0.002930   0.000000   0.012695   0.000000   0.000000   0.023438   

     texture63  texture64  
0     0.000000   0.025391  
1     0.039062   0.022461  
2     0.020508   0.002930  
3     0.000000   0.047852  
4     0.000000   0.031250  
..         ...        ...  
985   0.000000   0.018555  
986   0.000000   0.021484  
987   0.000000   0.001953  
988   0.001953   0.002930  
989   0.025391   0.022461  

[990 rows x 194 columns]

In [115]:
test

id   margin1   margin2   margin3   margin4   margin5   margin6  \
0       4  0.019531  0.009766  0.078125  0.011719  0.003906  0.015625   
1       7  0.007812  0.005859  0.064453  0.009766  0.003906  0.013672   
2       9  0.000000  0.000000  0.001953  0.021484  0.041016  0.000000   
3      12  0.000000  0.000000  0.009766  0.011719  0.017578  0.000000   
4      13  0.001953  0.000000  0.015625  0.009766  0.039062  0.000000   
..    ...       ...       ...       ...       ...       ...       ...   
589  1576  0.000000  0.000000  0.003906  0.015625  0.041016  0.000000   
590  1577  0.000000  0.003906  0.003906  0.005859  0.017578  0.000000   
591  1579  0.017578  0.029297  0.015625  0.013672  0.003906  0.015625   
592  1580  0.013672  0.009766  0.060547  0.025391  0.035156  0.025391   
593  1583  0.000000  0.117190  0.000000  0.019531  0.000000  0.136720   

      margin7   margin8   margin9  ...  texture55  texture56  texture57  \
0    0.005859  0.000000  0.005859  ...   0.006836   0.000000   0.015625   
1    0.007812  0.000000  0.033203  ...   0.000000   0.000000   0.006836   
2    0.023438  0.000000  0.011719  ...   0.128910   0.000000   0.000977   
3    0.003906  0.000000  0.003906  ...   0.012695   0.015625   0.002930   
4    0.009766  0.000000  0.005859  ...   0.000000   0.042969   0.016602   
..        ...       ...       ...  ...        ...        ...        ...   
589  0.017578  0.000000  0.005859  ...   0.098633   0.000000   0.004883   
590  0.017578  0.005859  0.000000  ...   0.012695   0.004883   0.004883   
591  0.025391  0.000000  0.000000  ...   0.073242   0.000000   0.028320   
592  0.039062  0.000000  0.003906  ...   0.003906   0.000000   0.000977   
593  0.001953  0.005859  0.000000  ...   0.107420   0.012695   0.016602   

     texture58  texture59  texture60  texture61  texture62  texture63  \
0     0.000977   0.015625        0.0        0.0   0.000000   0.003906   
1     0.001953   0.013672        0.0        0.0   0.000977   0.037109   
2     0.000000   0.000000        0.0        0.0   0.015625   0.000000   
3     0.036133   0.013672        0.0        0.0   0.089844   0.000000   
4     0.010742   0.041016        0.0        0.0   0.007812   0.009766   
..         ...        ...        ...        ...        ...        ...   
589   0.000000   0.003906        0.0        0.0   0.018555   0.000000   
590   0.002930   0.009766        0.0        0.0   0.090820   0.000000   
591   0.000000   0.001953        0.0        0.0   0.000000   0.042969   
592   0.000000   0.011719        0.0        0.0   0.000000   0.011719   
593   0.000977   0.004883        0.0        0.0   0.015625   0.000000   

     texture64  
0     0.053711  
1     0.044922  
2     0.000000  
3     0.008789  
4     0.007812  
..         ...  
589   0.000977  
590   0.016602  
591   0.006836  
592   0.018555  
593   0.017578  

[594 rows x 193 columns]

Encoded string to numerical, save train_columns and test_id. Then drop all for classification


In [116]:
le = LabelEncoder().fit(train.species)
labels = le.transform(train.species) 
print(labels)
classes = list(le.classes_)  
test_ids = test.id
train = train.drop(['species', 'id'], axis=1)  
test = test.drop(['id'], axis=1)

[ 3 49 65 94 84 40 54 78 53 89 98 16 74 50 58 31 43  4 75 44 83 84 13 66
 15  6 73 22 73 31 36 27 94 88 12 28 21 25 20 60 84 65 69 58 23 76 18 52
 54  9 48 47 64 81 83 36 58 21 81 20 62 88 34 92 79 82 20 32  4 84 36 35
 72 60 71 72 52 50 54 11 51 18 47  5  8 37 97 20 33  1 59  1 56  1  9 57
 20 79 29 16 32 54 93 10 46 59 84 76 15 10 15  0 69  4 51 51 94 36 39 62
  2 24 26 35 25 87  0 55 34 38  1 45  7 93 56 38 21 51 75 81 74 33 20 37
  9 40 60 31 83 50 71 67 30 66  1 43 61 23 65 84 87 46 57 16  2 28 12 96
 44 76 29 75 41 87 67 61 30  5 12 62  3 83 81  6 85  4 37 57 84 39 71 61
  6 76 14 31 98 40 17 51 16 42 63 86 37 69 86 71 80 78 14 35 25  5 39  8
  9 26 44 60 13 14 77 13 80 87 18 60 78 92 51 45 78 41 51 30 14 35 46 21
  8  6 92 38 40 15 32 17 93 71 92 27 78 15 19 60 21 38 36 49 74 67 95 31
 82 45 16 83 63 80 42 22 74 53 15 44 47 57 94 76 17 32 24 15 93 24 80 59
 46 12 51 77 79 70 69 16  2 63 83 55 12 53  1 67  0  2 36 42 10  9 52 59
  6 22 86 31 51 37 43 75 90 24 86 96 45 32 98 36 66

In [117]:
train.head()

margin1   margin2   margin3   margin4   margin5   margin6   margin7  \
0  0.007812  0.023438  0.023438  0.003906  0.011719  0.009766  0.027344   
1  0.005859  0.000000  0.031250  0.015625  0.025391  0.001953  0.019531   
2  0.005859  0.009766  0.019531  0.007812  0.003906  0.005859  0.068359   
3  0.000000  0.003906  0.023438  0.005859  0.021484  0.019531  0.023438   
4  0.005859  0.003906  0.048828  0.009766  0.013672  0.015625  0.005859   

   margin8   margin9  margin10  ...  texture55  texture56  texture57  \
0      0.0  0.001953  0.033203  ...   0.007812   0.000000   0.002930   
1      0.0  0.000000  0.007812  ...   0.000977   0.000000   0.000000   
2      0.0  0.000000  0.044922  ...   0.154300   0.000000   0.005859   
3      0.0  0.013672  0.017578  ...   0.000000   0.000977   0.000000   
4      0.0  0.000000  0.005859  ...   0.096680   0.000000   0.021484   

   texture58  texture59  texture60  texture61  texture62  texture63  texture64  
0   0.002930   0.035156        0.0        0.0   0.004883   0.000000   0.025391  
1   0.000977   0.023438        0.0        0.0   0.000977   0.039062   0.022461  
2   0.000977   0.007812        0.0        0.0   0.000000   0.020508   0.002930  
3   0.000000   0.020508        0.0        0.0   0.017578   0.000000   0.047852  
4   0.000000   0.000000        0.0        0.0   0.000000   0.000000   0.031250  

[5 rows x 192 columns]

In [118]:
test.head()

margin1   margin2   margin3   margin4   margin5   margin6   margin7  \
0  0.019531  0.009766  0.078125  0.011719  0.003906  0.015625  0.005859   
1  0.007812  0.005859  0.064453  0.009766  0.003906  0.013672  0.007812   
2  0.000000  0.000000  0.001953  0.021484  0.041016  0.000000  0.023438   
3  0.000000  0.000000  0.009766  0.011719  0.017578  0.000000  0.003906   
4  0.001953  0.000000  0.015625  0.009766  0.039062  0.000000  0.009766   

   margin8   margin9  margin10  ...  texture55  texture56  texture57  \
0      0.0  0.005859  0.023438  ...   0.006836   0.000000   0.015625   
1      0.0  0.033203  0.023438  ...   0.000000   0.000000   0.006836   
2      0.0  0.011719  0.005859  ...   0.128910   0.000000   0.000977   
3      0.0  0.003906  0.001953  ...   0.012695   0.015625   0.002930   
4      0.0  0.005859  0.000000  ...   0.000000   0.042969   0.016602   

   texture58  texture59  texture60  texture61  texture62  texture63  texture64  
0   0.000977   0.015625        0.0        0.0   0.000000   0.003906   0.053711  
1   0.001953   0.013672        0.0        0.0   0.000977   0.037109   0.044922  
2   0.000000   0.000000        0.0        0.0   0.015625   0.000000   0.000000  
3   0.036133   0.013672        0.0        0.0   0.089844   0.000000   0.008789  
4   0.010742   0.041016        0.0        0.0   0.007812   0.009766   0.007812  

[5 rows x 192 columns]

## Preprocessing


Splitting data:


Stratification is necessary for this dataset because there is a relatively large number of classes (100 classes for 990 samples). This will ensure we have all classes represented in both the train and test indices


In [119]:
sss = StratifiedShuffleSplit(n_splits=10, test_size=0.2)
sss.get_n_splits(train, labels)
for train_index, test_index in sss.split(train,labels):
    X_train, X_test = train.values[train_index], train.values[test_index]
    y_train, y_test = labels[train_index], labels[test_index]

## Classification:


### Decision Tree Classifier


In [120]:
from sklearn.tree import DecisionTreeClassifier

In [121]:
dtc = DecisionTreeClassifier()
dtc.fit(X_train, y_train)
train_predictions = dtc.predict(X_test)
print(train_predictions)
print(y_test)

[40 93 77 64  8  2 11 54 22 38 25  0 73 60 75 87 29 43 12 83 83 35 16 59
 76 42 13 65 73 38 80 31 59 19 28 48 80 92 83 44 20 29 14 80 84 88 39 92
 14 11 11  9 77 13 41 96 56 83 23 64 31 28 30  2 70 63 36 47 91 39 21 15
 66 97  9 70 69 52 89 91 72 74 17 33 84 77  4  0 78 32 65 69 10  8 95 71
  6 21 41 46 98 22  0 11 51 12 14 17  7 27 44 88 68 57  9 84 76 26 34 89
 12 93 76 98 47 78 97  7 62 76 90  8 83 10 85 63 12 37 82 52 41 61 46 45
 10 32 60 58 15 86 75 35 63 30 49 35 54 78 79 67 32 33 70 74  6 39 34 43
 36 61 90 45  8 94 39 23 55 24 51 77  2 86 60 67 87 82 19 24 96 18 68  1
 67 90 16 18 57 50]
[69 93 66 64 12 40  1 33 22 38 25  0 73 60 75 87 29 43  7 27 56  4 16 59
 76 42 13 65 73 38 80  5 59 19 28 48 71 92 83 44 20 22 14 80 84 88 39 92
 14 48 11  9 44 13 26 96 56 51 23 64 31 29 30  2 54 63 36 47 91 95 21 15
 66 97  4 70 79 52 89 91 72 74 17 33 84 77 31 20 78 32 65 96 49  8 95 71
  6 21 41 46 98 25  0 49 50 12 11 17  3 27 83 88 68 57  9  5 76 26 34 89
 53 93 37 98 47 72 97 85 62 40 

In [122]:
acc = accuracy_score(y_test, train_predictions)
print("Accuracy: {:.4%}".format(acc))

Accuracy: 71.7172%


### KNeighborsClassifier


In [123]:
from sklearn.neighbors import KNeighborsClassifier

In [124]:
knc = KNeighborsClassifier(5)
knc.fit(X_train, y_train)
train_predictions = knc.predict(X_test)
print(train_predictions)
print(y_test)

[69 93 77 64 12 40  1 33 67 38 25  0 73 60 75 87 29 43  7 27 56  4 16 59
 76 69 13 65 73 38 80 35 59 19 36 48 71 92 52 44 20 22 14 71 84 88 39 92
 14 41 11  9 44 13 19 96 53 51 23 64 31 29 30  2 54 63 22 19 91 11 21 15
 66 97  4 70 79 52 95 91 72 74 17 33 84 77 31 20 78 32 65 96 49  8 95 71
  6 73 12 46  1 25  0 49 50 12 11 57  3 27 13 88 68 44  9  5 76 17 34 89
 53 93 37 54 47 72 97 85 62 40 90 51 68 92 85 82  7 37 82 52 12 61 46 45
 92 42 60 58 15 86 75 35 63 30  3 35 53 78 81 70 32 29 54 74  6 94 40 56
 96 61 90 79  8 94 39 23 55 24 51 77  2 86 62 67 87 58 19 24 43 18 73  1
 67 69 16 18 57 56]
[69 93 66 64 12 40  1 33 22 38 25  0 73 60 75 87 29 43  7 27 56  4 16 59
 76 42 13 65 73 38 80  5 59 19 28 48 71 92 83 44 20 22 14 80 84 88 39 92
 14 48 11  9 44 13 26 96 56 51 23 64 31 29 30  2 54 63 36 47 91 95 21 15
 66 97  4 70 79 52 89 91 72 74 17 33 84 77 31 20 78 32 65 96 49  8 95 71
  6 21 41 46 98 25  0 49 50 12 11 17  3 27 83 88 68 57  9  5 76 26 34 89
 53 93 37 98 47 72 97 85 62 40 

In [125]:
acc = accuracy_score(y_test, train_predictions)
print("Accuracy: {:.4%}".format(acc))

Accuracy: 83.3333%


### RandomForestClassifier


In [126]:
from sklearn.ensemble import RandomForestClassifier

In [127]:
rfc = RandomForestClassifier()
rfc.fit(X_train, y_train)
train_predictions = rfc.predict(X_test)
print(train_predictions)
print(y_test)

[69 93 66 64 12 40  1 33 22 38 25  0 73 60 75 87 29 43  7 27 56  4 16 59
 76 42 13 65 73 38 80  5 59 19 28 48 71 92 83 44 20 22 14 80 84 88 39 92
 14 48 11  9 44 13 26 96 56 51 23 64 31 29 30  2 54 63 36 47 91 95 21 15
 66 97  4 70 79 52 89 91 72 74 17 33 84 77 31 20 78 32 65 96 10  8 95 71
  6 21 41 46 98 25  0 49 50 12 11 17  3 27 83 88 68 57  9  5 76 26 34 89
 53 93 37 98 47 72 97 85 62 40 90 79 27 10 85 82  7 37 82 52 41 61 46 45
 10 42 60 58 15 86 75 35 63 30  3 35 53 78 81 70 32 28 54 74  6 94 34 53
 36 61 90 45  8 94 39 23 55 24 51 77  2 86 62 67 87 58 19 24 43 18 81  1
 67 69 16 18 57 50]
[69 93 66 64 12 40  1 33 22 38 25  0 73 60 75 87 29 43  7 27 56  4 16 59
 76 42 13 65 73 38 80  5 59 19 28 48 71 92 83 44 20 22 14 80 84 88 39 92
 14 48 11  9 44 13 26 96 56 51 23 64 31 29 30  2 54 63 36 47 91 95 21 15
 66 97  4 70 79 52 89 91 72 74 17 33 84 77 31 20 78 32 65 96 49  8 95 71
  6 21 41 46 98 25  0 49 50 12 11 17  3 27 83 88 68 57  9  5 76 26 34 89
 53 93 37 98 47 72 97 85 62 40 

In [128]:
acc = accuracy_score(y_test, train_predictions)
print("Accuracy: {:.4%}".format(acc))

Accuracy: 98.4848%


### SVC


In [129]:
from sklearn.svm import SVC

In [130]:
svc = SVC()
svc.fit(X_train, y_train)
train_predictions = svc.predict(X_test)
print(train_predictions)
print(y_test)

[69 93 62 64 41 40  1 33 67 38 25  0 73 60 75 56 29 43  7 27 56  4 16 59
 76 30 13 65 73 38 80  2 59 19 28 48 71 92 52 44 20 22 14 80 84 88 39 92
 14 41 11  9 44 13 26 96 56 51 23 64 31 29 30  2 54 63 22 47 91 11 21 15
 66 97  4 70 79 52 95 91 72 74 17 33 84 77 31 20 78 32 65 96 49  8 95 71
  6 73 41 46 98 25  0 49 50 12 11 17  3 27 83 88 68 44  9  5 76 17 34 89
 53 93 37 54 47 72 97 85 62 40 90 79 68 92 85 82  7 37 82 52 41 61 46 45
 92 42 60 58 15 86 75 35 63 30  3 35 53 78 81 72 32 28 54 74  6 94 40 56
 36 61 90 45  8 94 39 23 55 24 51 77  2 86 62 67 87 58 19 24 43 18 73  1
 67 69 16 18 57 56]
[69 93 66 64 12 40  1 33 22 38 25  0 73 60 75 87 29 43  7 27 56  4 16 59
 76 42 13 65 73 38 80  5 59 19 28 48 71 92 83 44 20 22 14 80 84 88 39 92
 14 48 11  9 44 13 26 96 56 51 23 64 31 29 30  2 54 63 36 47 91 95 21 15
 66 97  4 70 79 52 89 91 72 74 17 33 84 77 31 20 78 32 65 96 49  8 95 71
  6 21 41 46 98 25  0 49 50 12 11 17  3 27 83 88 68 57  9  5 76 26 34 89
 53 93 37 98 47 72 97 85 62 40 

In [131]:
acc = accuracy_score(y_test, train_predictions)
print("Accuracy: {:.4%}".format(acc))

Accuracy: 88.8889%


### GaussianNB


In [132]:
from sklearn.naive_bayes import GaussianNB

In [133]:
gnb = GaussianNB()
gnb.fit(X_train, y_train)
train_predictions = gnb.predict(X_test)
print(train_predictions)
print(y_test)

[69 93 66 64 47 40  1 33 67 38 78  0 73 97 75 87 29 96  7 81 56  4 16 59
 85 90 13 65 73 70 80  5 59 19 22 48 71 26 83 44 20 22 14 80 84 88 40 92
 14 41 11  9 17 14 26 43 56 51 96 64 31 79 69 45 20 19 36 47 91 41 21 15
 66 97  4 70 79 52 89 91 36 74 17 33 84 19 31 20 78 32 78 85 26 19 95 80
  6 73 50 46  0 36  0 14 50 12 11 17  3 73 14 88 68 44  9  5 76 26 69 89
 26 93 37 98 47 19 50 85 91 36 69 79 50 10 85 19 13 39 82 52 41 54 46 47
 73 42 79 58 19 86 75  1 63 69  3 35 56 19 81 36 19 87 54 74  6 94 34 85
 36 85 90 69 11 41 70 36 55 69 51 77 69 86 62 67 87 58 19 85 69 18 73  1
 47 69 16 18 44 50]
[69 93 66 64 12 40  1 33 22 38 25  0 73 60 75 87 29 43  7 27 56  4 16 59
 76 42 13 65 73 38 80  5 59 19 28 48 71 92 83 44 20 22 14 80 84 88 39 92
 14 48 11  9 44 13 26 96 56 51 23 64 31 29 30  2 54 63 36 47 91 95 21 15
 66 97  4 70 79 52 89 91 72 74 17 33 84 77 31 20 78 32 65 96 49  8 95 71
  6 21 41 46 98 25  0 49 50 12 11 17  3 27 83 88 68 57  9  5 76 26 34 89
 53 93 37 98 47 72 97 85 62 40 

In [134]:
acc = accuracy_score(y_test, train_predictions)
print("Accuracy: {:.4%}".format(acc))

Accuracy: 62.1212%


## Finally


Finally, choose the one that have the most accuracy. In this case, we will choose RandomForestClassifier for building models
